DirectProgramming/C++SYCL_FPGA/Tutorials/Features/experimental/ip_authoring_mmhost/README.md

```diff
@@ -0,0 +1,381 @@
+### Memory-mapped interface parameters
+
+The following parameters are available for configuration:
+
+| Memory Attribute                 | Description               | Default Value
+---                                |---   |---
+| `buffer_location`          | The address space of the interface that associates with the host. | 1
+| `awidth`          | The width of the memory-mapped data bus in bits | 28
+| `dwidth`          | The width of the memory-mapped address bus in bits. | 64
+| `latency`         | The guaranteed latency from when a read command exits the component when the external memory returns valid read data. | 16
+| `readwrite_mode`  | The port direction of the interface. (0: Read & Write, 1: Read only, 2: Write only) | 0
+| `maxburst`        | The maximum number of data transfers that can associate with a read or write transaction. | 1
+| `align`           | The alignment of the base pointer address in bytes. | 0
+| `waitrequest`     | Adds the waitrequest signal that is asserted by the agent when it is unable to respond to a read or write request. | 0
+
+### Default interface
+
+If no memory-mapped interface parameters is informed for a pointer, then a mmhost interface with all default parameters is assumed for it.
+```c++
+struct PointerIP{
+  int *x, *y, *z;
+  int size;
+  
+  PointerIP(int *x_, int *y_, int *z_, int size_)
+    : x(x_), y(y_), z(z_), size(size_) {}
+
+    void operator()() const {
+      for (int i = 0; i < size; ++i) {
+        int mul = x[i] * y[i];
+        int add = mul + z[i];
+        y[i] = mul;
+        z[i] = add;
+      }
+    }
+};
+```
+
+### Memory-mapped interface implementation restriction
+
+Memory-mapped interface is restricted to function implementations, which means that it is not supported for lambda implementation. Also, all parameters must be preset, even if just one parameter is being changed from the default values.
+
+#### Example Functor
+```c++
+struct MyIP {
+  mmhost(
+    ... // All mmhost parameters.
+  ) int *my_pointer;
+
+  MyIP(int *input_pointer) : my_pointer(input_pointer) { ... } // Constructor for our IP.
+    ...
+  }
+
+  void operator()() const {
+    ...
+    // Functor code
+    my_pointer[x] = y;
+    ...
+  }
+}
+```
+
+
+#### Example 1: How to set-up a correct mmhost interface
+```c++
+// A memory-mapped interface must contain all parameters, in the following order:
+  mmhost(
+    1,       // buffer_location or aspace
+    28,      // address width
+    64,      // data width
+    16,      // latency
+    1,       // read_write_mode, 0: ReadWrite, 1: Read, 2: Write
+    1,       // maxburst
+    0,       // align, 0 defaults to alignment of the type
+    1        // waitrequest, 0: false, 1: true
+  ) int *memory_mapped_pointer;
+```
+
+#### Example 2: Changing default mmhost interface.
+If the mmhost argument is specified as a conduit interface then an input wire is generated at the top-level device image module with the same width as the address width property.
```

Comment:
This section needs cleanup and perhaps we need to add some info at the start of the tutorial too to set the stage:

An Avalon bus is always generated regardless of whether it is a conduit mmhost or a register_map mmhost. As many Avalon buses are generated as there are unique global memories in the system. The mmhost macros allow the user to specify unique memories and their properties.


For conduit mmhost arguments, a wire is created at the top level device image IP to carry the pointer address into the kernel. For register_map mmhost arguments, the input pointer is written into a register map that is associated with the kernel. In such cases, another Avalon bus is created for writing those arguments into a register map. Hence no separate wire is needed at the top level RTL module to input the address.

See the two regtests we have that show this difference:
https://github.com/intel-innersource/applications.fpga.oneapi.products.acl-test/blob/main/sycl/ip_authoring_flow/designs/vector_add_mmhost_register_map/prev/compile/ipa/reg.rout
https://github.com/intel-innersource/applications.fpga.oneapi.products.acl-test/blob/main/sycl/ip_authoring_flow/designs/vector_add_mmhost/prev/compile/ipa/reg.rout

Reply:
I removed some of the original description, and made a slightly newer one based on this description. Any suggestions to that or does more need to be added ?